File: python/src/aoc_2023_01.py
"""Advent of Code 2023 - Day 1."""

from doctest import testmod
from sys import stdin
from typing import Final


def calibration(s: str) -> int:
    """Return first and last digit on the line as an integer

    >>> calibration("pqr3stu8vwx")
    38
    >>> calibration("tree7the")
    77
    """
    digits: list[int] = [int(c) for c in s if c.isdigit()]
    return digits[0] * 10 + digits[-1]


WORD_DIGITS: Final[list[str]] = [
    "one",
    "two",
    "three",
    "four",
    "five",
    "six",
    "seven",
    "eight",
    "nine",
]


def starts_num_word(s: str) -> int | None:
    """If string starts with a digit or its word version, return it.

    >>> starts_num_word("2qw")
    2
    >>> starts_num_word("eight23")
    8
    >>> starts_num_word("a2") is None
    True
    """
    if s[0].isdigit():
        return int(s[0])
    for i, w in zip(range(1, 10), WORD_DIGITS, strict=True):
        if s.startswith(w):
            return i
    return None


def calibration_word(s: str) -> int:
    """Return first and last digit on the line as an integer allowing word digits.

    >>> calibration_word("two1nine")
    29
    >>> calibration_word("7sixteen")
    76
    """
    digits: list[int] = [
        x for x in [starts_num_word(s[i:]) for i in range(0, len(s))] if x is not None
    ]
    return digits[0] * 10 + digits[-1]


if __name__ == "__main__":
    testmod()
    lines: list[str] = list(stdin.readlines())
    print(sum(calibration(s) for s in lines))
    print(sum(calibration_word(s) for s in lines))

File: rust/src/bin/aoc_2023_01.rs
// Advent of Code, 2023 day 01

use aoc_rust::stdin_lines;
use itertools::enumerate;

// Convert string to a list of its digits
fn to_digits(s: &str) -> Vec<u32> {
    s.chars().filter_map(|c| c.to_digit(10)).collect()
}

const WORD_DIGITS: [&str; 9] = [
    "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
];

// Convert string to a list of its digits, allowing word forms
fn to_digits_words(s: &str) -> Vec<u32> {
    let mut digits: Vec<u32> = Vec::new();
    for (i, c) in s.char_indices() {
        if let Some(d) = c.to_digit(10) {
            digits.push(d);
        } else {
            let suffix = s.get(i..).unwrap();
            for (digit_i, digit_s) in enumerate(WORD_DIGITS) {
                if suffix.starts_with(digit_s) {
                    digits.push((digit_i + 1).try_into().unwrap());
                    break;
                }
            }
        }
    }
    digits
}

fn end_digits(xs: &[u32]) -> u32 {
    xs[0] * 10 + xs[xs.len() - 1]
}

fn main() {
    let lines: Vec<String> = stdin_lines().collect();

    let part_a: u32 = lines
        .iter()
        .map(|s| to_digits(s))
        .map(|xs| end_digits(&xs))
        .sum();

    let part_b: u32 = lines
        .iter()
        .map(|s| to_digits_words(s))
        .map(|xs| end_digits(&xs))
        .sum();

    println!("{part_a}");
    println!("{part_b}");
}
